common/lib/xmodule/xmodule/word_cloud_module.py

```diff
@@ -279,7 +279,7 @@ def student_view(self, context):  # lint-amnesty, pylint: disable=unused-argumen
         Renders the output that a student will see.
         """
         fragment = Fragment()
-        fragment.add_content(self.system.render_template('word_cloud.html', {
+        fragment.add_content(self.system.descriptor_runtime.render_template('word_cloud.html', {
```

Comment:
I am not clear why we need to explicitly specify `descriptor_system`. Shouldn't the attribute be gettable on `system` itself via the CombinedSystem [getattr](https://github.com/edx/edx-platform/blob/d6202880ae43845d1ce6f6027f7e10a5f48a034a/common/lib/xmodule/xmodule/x_module.py#L2015)?

Reply:
Besides the `descriptor_runtime` will go away once we eliminate the `ModuleSystem`.

Reply:
Another thing worth checking. Is `render_template()` really needed before the `ModuleSystem` is initialized? And if yes, for what? Maybe instead we should be removing the attribute in the `DescriptorSystem` and adding a service at the time the `ModuleSystem` is initialized.

Reply:
@symbolist Did you mean something like this? 1c8b5065f4009338343dd9a69f6f8520a9620f14

I can see adding the `replace_urls`, `replace_course_urls`, and `replace_jump_to_urls` functions to this service, and wrapping the provided `render_template` function to call these, so they don't have to be called by the XBlocks explicitly anymore.

Reply:
Thanks. this looks like a more useful direction. A few comments:

Since the `render_template()` method is only for [rendering Mako templates](https://github.com/edx/edx-platform/blob/3152c14192dcccac79b7b7a76eba91b380eba78a/common/djangoapps/edxmako/shortcuts.py#L157-L182) and the others can be used with any template engine, we should keep them separate. For `render_template()`, moving it to a `MakoService` in `common/djangoapps/edxmako/services.py` would be good. We can tackle the others later.

> I can see adding the replace_urls, replace_course_urls, and replace_jump_to_urls functions to this service, and wrapping the provided render_template function to call these, so they don't have to be called by the XBlocks explicitly anymore.

How many places is this happening? Will it also not result in the content being processed twice (the second time via `block_wrappers`) in places where `render_template()` is called in a `student_view()`?

Also did you check if the `render_template()` method in the `DescriptorSystem` is redundant? We can do this in a separate PR but we should also get rid of all the code in [mako_module.py](https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/mako_module.py) since I do not think it is being used any more. In case `get_context()` is being used it can be copied over to the relevant XBlocks.

Reply:
@symbolist  Ok, e975541 does:

* move `render_template` to a new `MakoService`.
  And since [xblock `RuntimeShim`](https://github.com/edx/edx-platform/blob/3152c14192dcccac79b7b7a76eba91b380eba78a/openedx/core/djangoapps/xblock/runtime/shims.py#L154) also deprecates `render_template` in favor of using the [`ResourceLoader.render_mako_template`](https://github.com/edx/xblock-utils/blob/0a8589acc99ecc608fb8fe4f6ad862eaf2bbc3ae/xblockutils/resources.py#L68-L76), I've deprecated this service call too, and will replace the platform calls to `runtime.render_template` with this preferred usage.
* reverted changes involving `replace_*urls`; we can deal with that separately.

I'll create a separate PR to remove DescriptorSystem.render_template and `MakoDescriptorSystem` -- I'm pretty sure these are not being used.

Reply:
@pomegranited Thanks!

Hmm, can `ResourceLoader.render_mako_template()` be used for XBlocks that are in edx-platform? It does seem to look up the templates in the package versus `render_template()` which [looks at](https://github.com/edx/edx-platform/blob/0c164ad4c24d8ee2a5d5f0a35cd057f21581866d/cms/envs/common.py#L558) the directories specified in settings. If that is the case and we can move the templates around to make it work, that would be fantastic.

Reply:
Hmm.. I gave the `ResourceLoader` a try, and found there's a lot of dependent (and shared) templates that would need to be moved around too, e.g. `common/djangoapps/pipeline_mako/templates/static_content.html`, `cms/templates/widgets/metadata-only-edit.html`.

So will just stick with using the new MakoService instead.